pkg/controllers/management/k3sbasedupgrade/k3sbased_upgrade_handler.go

```diff
@@ -177,13 +210,33 @@ func (h *handler) deployK3sBasedUpgradeController(clusterName string, isK3s, isR
 			}
 		}
 
-		if app.Spec.ExternalID == latestVersionID {
-			// app is up to date, no action needed
+		desiredApp := app.DeepCopy()
```

Comment:
I'd move this DeepCopy to line 229 because you aren't setting sub values in app until then. You can reference values using `app.` before that